### ISSUE TITLE
Bug: Triangulation::load()/save() with large output files

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#12752]
We have a bug reported for ASPECT related to Triangulation::load/save that got introduced between 9.2 and 9.3 (see [1]). After a quick look into tria_base.cc, I am sure that we will run into issues when the number of cells and/or the total file size runs above 32bits. I have not tested this, but this is suspicious:

https://github.com/dealii/dealii/blob/c090a8c3f9241a68294a647de39125a5ad1d18a4/source/distributed/tria_base.cc#L1606-L1607

https://github.com/dealii/dealii/blob/c090a8c3f9241a68294a647de39125a5ad1d18a4/source/distributed/tria_base.cc#L1656

https://github.com/dealii/dealii/blob/c090a8c3f9241a68294a647de39125a5ad1d18a4/source/distributed/tria_base.cc#L1383

https://github.com/dealii/dealii/blob/c090a8c3f9241a68294a647de39125a5ad1d18a4/source/distributed/tria_base.cc#L1250

[1]: https://github.com/geodynamics/aspect/issues/4349

**Comment by marcfehling** [comment]
The code snippets you mentioned were already there in version 9.2.0, but at a different spot. Using the `v9.2.0` tag for example
https://github.com/dealii/dealii/blob/b42193f72751984d28e11b249e47e3aebf6685f2/source/distributed/tria.cc#L1989-L1990

The entire `DataTransfer` class has been moved from `p:d:Tria` to `p:DistributedTriangulationBase` to be used in combination with `p:fd:Tria`. `DataTransfer` has not been changed between releases AFAIK. I would bet that the error might be somewhere else, but I don't have an idea right now :(

**Comment by bangerth** [comment]
At least the first of these code snippets appears correct to me since it pertains to per-cell data. I agree that everything that invokes `global_num_cells` looks suspicious and should probably be `types::global_cell_index`.

**Comment by tjhei** [comment]
> At least the first of these code snippets appears correct to me since it pertains to per-cell data. I

I disagree. This is an offset into the data file and is used in the MPI IO call where the type is ``MPI_Offset`` (64 bit). If your file is bigger than 4 GB, this will fail.
